L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,359 @@
+
+      auto mapElem(tcToCalo.find(cell.detId()));
+      if (mapElem == tcToCalo.end())
+        continue;
+
+      outMap->insert(mapElem->second, edm::Ref<l1t::HGCalTriggerCellBxCollection>(triggerCellsHandle, triggerCells.key(cItr)));
+
+      if (makeCellsCollection_) {
+        auto& simEnergies(tcToEnergies.at(cell.detId()));
+        if (simEnergies.first > 0.) {
+          outCollection->push_back(bx, cell);
+          (*outCollection)[outCollection->size() - 1].setMipPt(cell.mipPt() * simEnergies.second / simEnergies.first);
+        }
+      }
+
+      triggerCellPtrs.emplace_back(triggerCellsHandle, triggerCells.key(cItr));
+    }
+  }
+
+  event.put(std::move(outMap));
+  if (makeCellsCollection_)
+    event.put(std::move(outCollection));
+
+  auto outClusters(std::make_unique<l1t::HGCalClusterBxCollection>());
+
+  auto sortCellPtrs([](TriggerCellPtr const& lhs, TriggerCellPtr const& rhs)->bool {
+      return lhs->mipPt() > rhs->mipPt();
+    });
+
+  std::sort(triggerCellPtrs.begin(), triggerCellPtrs.end(), sortCellPtrs);
+  dummyClustering_.clusterizeDummy(triggerCellPtrs, *outClusters);
+
+  std::unordered_map<unsigned, std::vector<unsigned>> caloToClusterIndices;
+  for (unsigned iC(0); iC != outClusters->size(); ++iC) {
+    auto const& cluster((*outClusters)[iC]);
+    // cluster detId and cell detId are identical
+    auto caloRef(tcToCalo.at(cluster.detId()));
+    caloToClusterIndices[caloRef.key()].push_back(iC);
+  }
+
+  auto&& clustersHandle(event.put(std::move(outClusters)));
+
+  auto outMulticlusters(std::make_unique<l1t::HGCalMulticlusterBxCollection>());
+
+  for (auto const& ocr : orderedCaloRefs) {
+    auto const& ref(ocr.second);
+    
+    if (ref.isNull()) // shouldn't happen
+      continue;
+
+    auto const& caloParticle(*ref);
+
+    l1t::HGCalMulticluster multicluster;
+    
+    for (unsigned iC : caloToClusterIndices[ref.key()]) {
+      ClusterPtr clPtr(clustersHandle, iC);
+      multicluster.addConstituent(clPtr, true, 1.);
+    }
+
+    // Set the gen particle index as the DetId
+    multicluster.setDetId(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
+
+    auto&& centre(multicluster.centre());
+    math::PtEtaPhiMLorentzVector multiclusterP4(multicluster.sumPt(), centre.eta(), centre.phi(), 0.);
+    multicluster.setP4(multiclusterP4);
+
+    showerShape_.fillShapes(multicluster, geometry);
+
+    // not setting the quality flag
+    // multicluster.setHwQual(id_->decision(multicluster));
+    // fill H/E
+    multicluster.saveHOverE();            
+    
+    outMulticlusters->push_back(0, multicluster);
+  }
+
+  event.put(std::move(outMulticlusters));
+}
+
+void
+CaloTruthCellsProducer::beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const& setup)
+{
+  // check geometry version
+
+  edm::ESHandle<CaloGeometry> geom;
+  setup.get<CaloGeometryRecord>().get(geom);
+  auto const* eegeom(geom->getSubdetectorGeometry(DetId::HGCalEE, ForwardSubdetector::ForwardEmpty));
```

Comment:
The V8/V9 check can be done directly with the trigger geometry:
https://github.com/PFCal-dev/cmssw/blob/hgc-tpg-devel-CMSSW_10_6_0_pre4/L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h#L31

Reply:
This is good to know. Incorporated.